### ISSUE TITLE
Component that will allows keeping all important tips in sync on startup. 

### ISSUE THREAD

**Issue by noescape00** [stratisproject/StratisBitcoinFullNode#2503]
### Problem

When node shuts down incorrectly (exception\reboot or any other problem that led to it) we need to recover the node next time we launch it. In order to do that we need to initialize tips of all important components at the same height. Components that have to be initialized on the same height at startup are: StakeChain, ChainRepository, CoinView, BlockStore, ProvenheadersRepository (after PH activation height and last checkpoint).

### Solution

We need to introduce a new component that keeps track of last hash and height that was persisted by all components that need to be in sync on startup.

Let's call this new component **TipsManager** (please feel free to suggest a better name for it).

Here is a suggested interface for it: 

```
public interface ITipsManager : IDisposable
{
    // Type is the type of component that requires syncing the tips.
    void RegisterTipProvider(Type type);

    void CommitTipPersited(Type type, ChainedHeader tip);

    // Returns highest tip commited by all components.
    HashHeightPair GetLastCommonTip();
}
```

Every feature will be able to register components on tipsManager during feature initialization.

Every feature that owns a component that requires tips syncing (StakeChain, ChainRepository, CoinView, BlockStore, ProvenheadersRepository) will override initialize method and call `ITipsManager.RegisterTipProvider` with an argument of component that requires tips syncing.

When any component is being initialized it should initialize at height equal to `ITipsManager.GetLastCommonTip().Height` and do a sanity check that hash of the tip is correct.

When any component advances and persists data to the database during node syncing it should call `ITipsManager.CommitTipPersited` AFTER data is persisted. We require every component to guarantee that once tip was commited the node can't loose data of that tip. ITipsManger should also support cases when components commit tip that is below last commited tip. 

`ITipsManager` keeps track of all tips commited and it has it's own tip which is always equal to the highest common commited tip. For example let's say we have 3 components that are registered: component A,B and C.

```
A commits height 5
B commits height 5
C commits height 3
last common tip is 3
C commits height 5
last common tip is 5
C commits height 4
last common tip is 4
C commits height 8
B commits height 9
A commits height 7
last common tip is 7
```

When last common tip changes `LastCommonTip` should be persisted in the background to the database by `ITipsManager`.

### Benefits

Components that require syncing don't need to be initialized in particular order, they don't need to be in the same namespace or have references to each other, they don't need to be aware of other components and know anything about how many there are.

**Comment by MithrilMan** [comment]
I like the idea of having a sync in place to ensure a common tip, I just need to understand this detail:

> When any component is being initialized it should initialize at height equal to ITipsManager.GetLastCommonTip().Height and do a sanity check that hash of the tip is correct.

If you register during initialization, and you check for common tip, wouldn't you receive different values depending the order of components initializations?
e.g. the first component that register and call it, will get its own height, the second will get the common between the 1st and the 2nd, and so forth

**Comment by bokobza** [comment]
@MithrilMan if I understand correctly, the common tip is saved to the DB so when component are initialized, they'll all get the same tip.

**Comment by bokobza** [comment]
> ITipsManager keeps track of all tips commited 

Where is this logic done? In `void CommitTipPersited`?

CommonTipManager? LastTipManager? LastCommonTipManager?

What happens when there is a reorg? Do we need to notify the manager?

**Comment by noescape00** [comment]
>If you register during initialization, and you check for common tip, wouldn't you receive different values >depending the order of components initializations?
>e.g. the first component that register and call it, will get its own height, the second will get the >common between the 1st and the 2nd, and so forth

common tip is loaded from db. 
so tip sources registration has nothing to do with the common tip- common tip comes from the db. 


so here is an example: 
we have components A,B,C. 
We run the node and TipsManager initializes with common tip == 0 because it's the first run. 
A,B,C initialize at TipsManager.LastCommonTip which is 0. 
Now we run the node for some time: 
A commits height 10
B commits height 12
C commits height 8
TipsManager saves 8 as the last common tip. 
Now A commits 20
node is being killed by PC reboot
____

now we start the node and we initialize A, then B, then C
A will go to TipsManager and it will load 8 from db. A will rewind from 20 to 8
B will rewind to 8
C will initialize at 8 and do nothing

Now all the components were initialized and they start commiting their heights as they advance

**Comment by noescape00** [comment]
>Where is this logic done? In void CommitTipPersited?

yeah. 
so in tips manager we have tips of all components and whenever new tip is claimed we check if the component that claimed new tip used to claim the lowest tip. if it did we check who is the lowest tip right now and if it's higher than prev value we update entry in db. 
we also probably should check that every single tip has ancestor of the fork point we are about to commit (to avoid thinking that all components at tip X but those are tips from different chains and we are in the middle of reorg or smth)

**Comment by noescape00** [comment]
>What happens when there is a reorg? Do we need to notify the manager?

Yeah. Every time we change a tip of every component we notify the manager. 
Last common tip can change it's height in both directions 

>Re names: CommonTipManager? LastTipManager? LastCommonTipManager?

I like CommonTipManager

**Comment by bokobza** [comment]
OK, concept ACK from me.

**Comment by dangershony** [comment]
Generally I don't like the word manager so much.
I would go for `TipSynchronizer` or `CommotTipStore`

**Comment by fassadlr** [comment]
Merged with #2814
